tests/cancun/eip4788_beacon_root/test_beacon_root_precompile.py

```diff
@@ -0,0 +1,243 @@
+
+    The expected result is that the precompile call will return the correct
+    `parent_beacon_block_root` for a valid input timestamp and return the zero'd 32 bytes value
+    for an invalid input timestamp.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "beacon_root, timestamp",
+    [
+        (12, 12),  # twelve
+        (2**32, 2**32),  # arbitrary
+        (2**64 - 2, 2**64 - 2),  # near-max
+        (2**64 - 1, 2**64 - 1),  # max
+    ],
+    indirect=["beacon_root"],
+)
+@pytest.mark.valid_from("Cancun")
+def test_beacon_root_equal_to_timestamp(
+    state_test: StateTestFiller,
+    env: Environment,
+    pre: Dict,
+    tx: Transaction,
+    post: Dict,
+):
+    """
+    Tests the beacon root precompile call where the beacon root is equal to the timestamp.
+
+    The expected result is that the precompile call will return the `parent_beacon_block_root`,
+    as all timestamps used are valid.
+    """
+    state_test(
+        env=env,
+        pre=pre,
+        txs=[tx],
+        post=post,
+    )
+
+
+@pytest.mark.parametrize(
+    "call_gas, valid_call",
+    [
+        (BEACON_ROOT_PRECOMPILE_GAS, True),
+        (BEACON_ROOT_PRECOMPILE_GAS + 1, True),
+        (BEACON_ROOT_PRECOMPILE_GAS - 1, False),
+    ],
+)
+@pytest.mark.parametrize(
+    "timestamp",
+    [
+        12,  # twelve
+        2**32,  # arbitrary
+        2**64 - 2,  # near-max
+        2**64 - 1,  # max
+    ],
+)
+@pytest.mark.valid_from("Cancun")
+def test_beacon_root_tx_timestamp_collisions(
+    state_test: StateTestFiller,
+    env: Environment,
+    pre: Dict,
+    tx: Transaction,
+    precompile_call_account: Account,
+    timestamp: int,
+    valid_call: bool,
+):
+    """
+    Tests multiple beacon root precompile calls where the timestamp index is calculated to
+    be equal for each call (i.e colliding). For each parameterized timestamp a list of
+    colliding timestamps are calcuted using factors of the `HISTORY_ROOTS_MODULUS`.
+
+    The expected result is that precompile call will return an equal `parent_beacon_block_root`
+    for each timestamp used within the call, as the timestamp index used will be the same.
+
+    Here we are predomintely testing that the `timestamp_index` and `root_index` are derived
+    correctly in the evm.
+    """
+    post = {}
+    txs, colliding_timestamps = [], []
+    timestamp_collisions = 5
+    for i in range(timestamp_collisions):
+        pre[to_address(0x100 + i)] = precompile_call_account
+        colliding_timestamps.append(timestamp + i * HISTORICAL_ROOTS_MODULUS)
+
+    # check timestamp_index function is working as expected
+    timestamp_indexes = [timestamp_index(v) for v in colliding_timestamps]
+    assert len(set(timestamp_indexes)) == 1, "timestamp_index function is not working"
+
+    for i, timestamp in enumerate(colliding_timestamps):
+        txs.append(
```

Comment:
Multiple transactions in a single state test doesn't follow the convention of single tx state test of `ethereum/tests`, so we might deprecate it in this repo too in the future, and this might need to be a BlockchainTest.